javaslang/src/main/java/javaslang/collection/PriorityQueue.java

```diff
@@ -544,210 +544,214 @@ public boolean equals(Object o) {
-         * * | insert (x, ts) =            Node (x,0,[]) :: ts
-         **/
-        static <T> Seq<Node<T>> insert(SerializableComparator<? super T> comparator, T element, Seq<Node<T>> forest) {
-            final Node<T> tree = Node.of(element, 0, List.empty());
-            if (forest.size() >= 2) {
-                final Seq<Node<T>> tail = forest.tail();
-                final Node<T> t1 = forest.head(), t2 = tail.head();
-                if (t1.rank == t2.rank) {
-                    return tree.skewLink(comparator, t1, t2).appendTo(tail.tail());
-                }
-            }
-            return tree.appendTo(forest);
-        }
+    /* Based on http://www.brics.dk/RS/96/37/BRICS-RS-96-37.pdf */
+    static final class Node<T> implements Serializable {
+
+        private static final long serialVersionUID = 1L;
 
-        /** fun meld (ts, ts') = meldUniq (uniqify ts, uniqify ts') */
-        static <T> Seq<Node<T>> meld(SerializableComparator<? super T> comparator, Seq<Node<T>> source, Seq<Node<T>> target) {
-            return meldUnique(comparator, uniqify(comparator, source), uniqify(comparator, target));
+        final T root;
+        final int rank;
+        final Seq<Node<T>> children;
+
+        private Node(T root, int rank, Seq<Node<T>> children) {
+            this.root = root;
+            this.rank = rank;
+            this.children = children;
+
+            assert children.forAll(c -> c.rank < this.rank);
         }
 
-        /**
-         * fun uniqify [] = []
-         * *  | uniqify (t :: ts) = ins (t, ts) (∗ eliminate initial duplicate ∗)
-         **/
-        static <T> Seq<Node<T>> uniqify(SerializableComparator<? super T> comparator, Seq<Node<T>> forest) {
-            return forest.isEmpty()
-                   ? forest
-                   : ins(comparator, forest.head(), forest.tail());
+        static <T> Node<T> of(T value, int rank, Seq<Node<T>> children) {
+            return new Node<>(value, rank, children);
         }
 
-        /**
-         * fun ins (t, []) = [t]
-         * * | ins (t, t' :: ts) = (∗ rank t ≤ rank t' ∗)
-         * *     if rank t < rank t' then t :: t' :: ts
-         * *     else                     ins (link (t, t'), ts)
+        /*
+         * fun link (t1 as Node (x1,r1,c1), t2 as Node (x2,r2,c2)) = (∗ r1 = r2 ∗)
+         * *  if Elem.leq (x1,x2) then Node (x1,r1+1,t2 :: c1)
+         * *  else                     Node (x2,r2+1,t1 :: c2
          */
-        static <T> Seq<Node<T>> ins(SerializableComparator<? super T> comparator, Node<T> tree, Seq<Node<T>> forest) {
-            while (!forest.isEmpty() && tree.rank == forest.head().rank) {
-                tree = tree.link(comparator, forest.head());
-                forest = forest.tail();
-            }
-            return tree.appendTo(forest);
+        Node<T> link(Comparator<? super T> comparator, Node<T> tree) {
+            assert rank == tree.rank;
+
+            return comparator.compare(this.root, tree.root) <= 0
```

Comment:
Yes, the SerializableComparator disappeared. 80% of the special methods (greater, greaterOrEqual, ...) weren't used at all - dead code. When you forget about the assert statements, there are only 2 uses of lessOrEqual() and 1 use of less(). It does not scale and is uncommon - an additional DSL abstraction layer above a well known and concise Java API. We don't need it here.